text/0000-engines-component-dependencies.md

```diff
@@ -0,0 +1,97 @@
+- Start Date: (fill me in with today's date, YYYY-MM-DD)
+- RFC PR: (leave this empty)
+- Ember Issue: (leave this empty)
+
+# Summary
+
+This RFC proposes to allow the declaration of components as dependencies of engines.
+
+# Motivation
+
+Currently, the only way to allow engines to share components with their consuming application/engine is to extract the component and package it as a standalone add-on. The add-on can then be installed by any application/engine that requires it.
+
+While this approach is a perfectly viable option in some circumstances, the viability begins to breakdown in distributed applications where the developer may not have control of some or all of the engines being consumed.
+
+In such circumstances it becomes difficult to ensure consistency throughout the application as each engine could theoretically be using a different version of the same component, potentially resulting in different areas of the application having different UI/UX for the same component/functionality.
+
+Allowing engines to declare components as dependencies would resolve this issue of consistency. If any changes were made to components in the main application, these changes would be able flow down through to each of the mounted engines without requiring the engine developers to change any of their code.
+
+# Detailed design
+
+As it is already possible to declare services as dependencies of engines, the underlying groundwork should already be in place to also allow components to be declared as dependencies. Declaring a component dependency would look and work exactly the same way as declaring service dependencies.
+
+Consider the following, using the `ember-blog` engine as an example:
+
+### Declaring Dependencies (Engine)
+
+As with services, the engine would be required to declare any components that they expected to be supplied as dependencies.
```

Comment:
> How are nested components handled? Would this require Engine authors to understand the internal composition of the components they depend on?

Do you think it would be unreasonable to expect Engine authors to understand the internal composition of the components that they're depending on?

A component would presumably have come from one of two places:

1. It's a custom component built by them, in which case they would already understand it.
2. From an add-on which would presumably have documentation for the Engine author to read.

> Edit: Similarly, what about service dependencies inside a component's JS?

I hadn't actually thought of this, what are your thoughts? I'm not 100% sure what is possible here.